### PR TITLE
Action Forms Refactor 6: New ActionForm Component

### PR DESCRIPTION
**Pull request by iethree**
The final 🤞  chapter.

> **Note**
> The diff is scary but half of it is a big new test spec

Partly resolves https://github.com/metabase/metabase/issues/26724

## Description

- Updates `FormCategoryInput` to work like all the other form inputs in our new forms framework
- Creates a new `ActionForm` component that encapsulates turning `ActionFormSettings` into `ActionFormFieldProps` and simplifies a bunch of other components
- Moves certain `utils` functions from `FormCreator/utils` to `ActionForm/utils`
- Adds unit tests!

## Testing

Implicit edit forms properly load field values and execute properly.
editFormsWork

Inline custom forms execute properly.
inlineformsWork

The form creator looks alright and updates form settings properly.
can Update things

### CODE REVIEW

File: frontend/src/metabase/components/form/FormWidget.jsx [hunk 1]
```diff
@@ -3,7 +3,6 @@ import React, { forwardRef } from "react";
 
 import { PLUGIN_FORM_WIDGETS } from "metabase/plugins";
 
-import CategoryFieldPicker from "metabase/core/components/FormCategoryInput";
```
Comment (iethree) [inline frontend/src/metabase/components/form/FormWidget.jsx:6]: https://github.com/metabase/metabase/pull/27169#discussion_r1046103376

File: frontend/src/metabase/components/form/FormWidget.jsx [hunk 2]
```diff
@@ -34,14 +33,12 @@ const WIDGETS = {
   section: FormSectionWidget,
   select: FormSelectWidget,
   integer: FormNumericInputWidget,
-  number: FormNumericInputWidget,
```
Comment (iethree) [inline frontend/src/metabase/components/form/FormWidget.jsx:37]: this was added in temporarily to keep PR's independently reviewable

File: frontend/src/metabase/actions/components/ActionForm/utils.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
+import { t } from "ttag";
```
Comment (iethree) [inline frontend/src/metabase/actions/components/ActionForm/utils.ts:1]: all of these functions are moved from FormCreator/utils

**Review by alxnddr — APPROVED** [review]
Looks good to me, although I'm still learning this codebase

File: frontend/src/metabase/actions/components/ActionCreator/FormCreator/utils.ts [hunk 4]
```diff
@@ -230,6 +144,9 @@ export const getInputType = (param: Parameter, field?: Field) => {
   if (field.isDate()) {
     return field.isDateWithoutTime() ? "date" : "datetime";
   }
+  if (field.semantic_type === TYPE.Email) {
+    return "string";
+  }
```
Comment (kulyk) [inline frontend/src/metabase/actions/components/ActionCreator/FormCreator/utils.ts:149]: This makes me think: don't we have an `email` input type? 🤔 
Also nit: maybe it'd be nicer to combine this with [this check](https://github.com/metabase/metabase/blob/98182eab7179f8d24acc06b75f4620b81c6d4028/frontend/src/metabase/actions/components/ActionCreator/FormCreator/utils.ts#L157) below

File: frontend/src/metabase/actions/components/ActionCreator/FormCreator/FormCreator.tsx [hunk 5]
```diff
@@ -56,103 +40,29 @@ export function FormCreator({
...
-              )}
-            </div>
-          )}
-        </Droppable>
-      </DragDropContext>
+      <ActionForm
+        parameters={sortedParams}
+        initialValues={[]}
+        onClose={_.noop}
+        onSubmit={_.noop}
```
Comment (kulyk) [inline frontend/src/metabase/actions/components/ActionCreator/FormCreator/FormCreator.tsx:62]: Why do we pass `noop` here?

File: frontend/src/metabase/actions/components/ActionCreator/FormCreator/FormCreator.tsx [hunk 6]
```diff
@@ -56,103 +40,29 @@ export function FormCreator({
...
-              {!params.length && (
-                <EmptyFormPlaceholder onExampleClick={onExampleClick} />
-              )}
-            </div>
-          )}
-        </Droppable>
-      </DragDropContext>
+      <ActionForm
+        parameters={sortedParams}
+        initialValues={[]}
```
Comment (kulyk) [inline frontend/src/metabase/actions/components/ActionCreator/FormCreator/FormCreator.tsx:60]: Isn't `initialValues` supposed to be an object?

File: frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+import type {
+  ActionFormSettings,
+  FieldSettings,
+  WritebackParameter,
+} from "metabase-types/api";
+
+import type { Parameter } from "metabase-types/types/Parameter";
+
+import { reorderFields } from "metabase/actions/components/ActionCreator/FormCreator";
+import Icon from "metabase/components/Icon";
```
Comment (kulyk) [inline frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx:27]: Think this import should go right after `metabase/core/components` imports

File: frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+  FormFieldContainer,
+  SettingsContainer,
+  InputContainer,
+} from "./ActionForm.styled";
+
+import { getForm } from "./utils";
+
+export interface ActionFormComponentProps {
+  parameters: WritebackParameter[] | Parameter[];
+  initialValues?: any;
```
Comment (kulyk) [inline frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx:41]: Can we avoid `any` here?

File: frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+  onClose?: () => void;
+  onSubmit: (params: any, actions: any) => void;
```
Comment (kulyk) [inline frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx:43]: Cam we avoid `any` here?

File: frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+  initialValues = {},
+  onClose,
+  onSubmit,
+  submitTitle,
+  submitButtonColor,
+  formSettings,
+  setFormSettings,
+}: ActionFormComponentProps): JSX.Element => {
+  // allow us to change the color of the submit button
+  const submitButtonVariant = { [submitButtonColor ?? "primary"]: true };
```
Comment (kulyk) [inline frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx:61]: Should we just make `primary` a default `submitButtonColor` value?

File: frontend/src/metabase/actions/components/ActionForm/ActionForm.unit.spec.tsx [hunk 11]
```diff
@@ -0,0 +1,504 @@
...
+): Partial<WritebackParameter> => {
+  const id = overrides?.id ?? "abc-123";
+  return {
+    id,
+    required: false,
+    target: ["variable", ["template-tag", id]],
+    type: "type/Text",
+    ...overrides,
+  };
+};
```
Comment (kulyk) [inline frontend/src/metabase/actions/components/ActionForm/ActionForm.unit.spec.tsx:42]: Does it make sense to extract these to `metabase-types/api/mocks`?

File: frontend/src/metabase/core/components/FormCategoryInput/CategoryFieldInput/CategoryFieldInput.tsx [hunk 12]
```diff
@@ -31,22 +31,22 @@ const DefaultTokenFieldLayout = ({
...
   onChange,
 }: CategoryFieldInputProps) {
   return (
     <FieldValuesWidgetContainer>
       <StyledFieldValuesWidget
         // eslint-disable-next-line @typescript-eslint/ban-ts-comment
         // @ts-ignore, typescript is very confused about the value type for this non-tsx component
-        value={value ? [String(value)] : ([] as any)}
-        fields={[field]}
+        value={value ? [String(value)] : ([""] as any)}
```
Comment (kulyk) [inline frontend/src/metabase/core/components/FormCategoryInput/CategoryFieldInput/CategoryFieldInput.tsx:48]: This expression feels a bit too complicated. Is there anything we can do to make it easier to read?

**Review by kulyk — APPROVED** [review]
Awesome work! Left a few ideas

File: frontend/src/metabase/actions/components/ActionCreator/FormCreator/utils.ts (hunk 4 above)
Comment (iethree) [reply frontend/src/metabase/actions/components/ActionCreator/FormCreator/utils.ts:149]: there is an email input component, but we don't allow users to set email fields in the form creator, so i was trying to align this with the available options.

File: frontend/src/metabase/actions/components/ActionCreator/FormCreator/FormCreator.tsx (hunk 5 above)
Comment (iethree) [reply frontend/src/metabase/actions/components/ActionCreator/FormCreator/FormCreator.tsx:62]: Until we tackle https://github.com/metabase/metabase/issues/26701, we don't even show the submit/cancel buttons in the form creator.

It's a little easier type-wise to pass `_.noop` than to make these functions optional.

File: frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx (hunk 7 above)
Comment (iethree) [reply frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx:27]: thanks. we gotta figure out how to make the linter do this 🤔

File: frontend/src/metabase/actions/components/ActionForm/ActionForm.unit.spec.tsx (hunk 11 above)
Comment (iethree) [reply frontend/src/metabase/actions/components/ActionForm/ActionForm.unit.spec.tsx:42]: 😬 I fear that making these general purpose would require a good bit more work and thinking, and would also make the tests more complex. I think (and I could be wrong about this), that these quick little mocks work best when they don't try to serve more than a few tests.

File: frontend/src/metabase/core/components/FormCategoryInput/CategoryFieldInput/CategoryFieldInput.tsx (hunk 12 above)
Comment (iethree) [reply frontend/src/metabase/core/components/FormCategoryInput/CategoryFieldInput/CategoryFieldInput.tsx:48]: 🤦  yes. 😳 

```
value={[String(value ?? '')]}
```

**Pull request merged**